nav2_loopback_sim/src/loopback_simulator.cpp

```diff
@@ -0,0 +1,227 @@
+  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
+}
+
+LoopbackSimulator::~LoopbackSimulator()
+{
+  RCLCPP_INFO(get_logger(), "Destroying %s", get_name());
+}
+
+
+void LoopbackSimulator::publishTransform(
+  const geometry_msgs::msg::Pose & pose,
+  const std::string & parent_frame,
+  const std::string & child_frame)
+{
+  geometry_msgs::msg::TransformStamped transformStamped;
+  transformStamped.header.stamp = this->now() + rclcpp::Duration::from_seconds(dt_);
+  transformStamped.header.frame_id = parent_frame;
+  transformStamped.child_frame_id = child_frame;
+  transformStamped.transform.translation.x = pose.position.x;
+  transformStamped.transform.translation.y = pose.position.y;
+  transformStamped.transform.translation.z = pose.position.z;
+  transformStamped.transform.rotation = pose.orientation;
+
+  tf_broadcaster_->sendTransform(transformStamped);
+}
+
+void LoopbackSimulator::twistCallback(const geometry_msgs::msg::Twist::SharedPtr msg)
+{
+  if (!init_pose_set_) {
+    RCLCPP_WARN(
+      rclcpp::get_logger("logger_name"),
+      "Initial pose has not been set. Ignoring incoming velocity messages.");
+    return;
+  }
+
+  // get base_link location in the odom frame
+  geometry_msgs::msg::TransformStamped transformStamped;
+  try {
+    transformStamped = tf_buffer_->lookupTransform("odom", "base_link", tf2::TimePointZero);
+  } catch (tf2::TransformException & ex) {
+    RCLCPP_ERROR(this->get_logger(), "%s", ex.what());
+    return;
+  }
+
+  // Convert the quaternion in the transform to tf2::Quaternion
+  tf2::Quaternion current_orientation;
+  tf2::fromMsg(transformStamped.transform.rotation, current_orientation);
+
+  // Calculate the change in orientation based on the twist message
+  tf2::Quaternion angular_change;
+  angular_change.setRPY(msg->angular.x * dt_, msg->angular.y * dt_, msg->angular.z * dt_);
+
+  // Integrate the change in orientation with the current orientation
+  tf2::Quaternion new_orientation = current_orientation * angular_change;
+
+  // Calculate the change in position based on the twist message
+  tf2::Matrix3x3 rotation_matrix(new_orientation);
+  tf2::Vector3 linear_velocity(msg->linear.x, msg->linear.y, msg->linear.z);
+  tf2::Vector3 rotated_linear_velocity = rotation_matrix * linear_velocity;
+
+  // Integrate the change in position with the current position
+  geometry_msgs::msg::Pose new_position;
+  new_position.position.x =
+    transformStamped.transform.translation.x + rotated_linear_velocity.x() * dt_;
+  new_position.position.y =
+    transformStamped.transform.translation.y + rotated_linear_velocity.y() * dt_;
+  new_position.position.z =
+    transformStamped.transform.translation.z + rotated_linear_velocity.z() * dt_;
+  new_position.orientation = tf2::toMsg(new_orientation);
+
+  // Publish the updated transform between odom and base_link
+  publishTransform(new_position, "odom", "base_link");
```

Comment:
This function seems overcomplicated. Why not use TF math more to do the rotations and matrix calculation for you?